### PR TITLE
Improved S-FEEL expressions evaluation performance. Resolved #201

### LINKED ISSUE #201: Improve feel evaluation performance
Currently FEEL evaluation is sub standard.

### PR DESCRIPTION
**Pull request by stlhrt**

### CODE REVIEW

File: src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java [hunk 1]
```diff
@@ -88,38 +94,71 @@ public Serializable evaluateOutputEntry(final OutputEntry outputEntry, final Eva
...
             throw new ExpressionEvaluationException("Can not evaluate feel expression '" + expression.getValue() + "'", e);
         }
 
         return result;
     }
+
+    /**
+     * {@link CachingMapVariableResolverFactory} uses exception driven programming instead of null checks.
+     */
+    private static class FastCachingMapVariableResolverFactory extends CachingMapVariableResolverFactory {
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java]: Can you consider to move to separated file? Meybe would be desirable to create mvel package?

File: src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java [hunk 2]
```diff
@@ -88,38 +94,71 @@ public Serializable evaluateOutputEntry(final OutputEntry outputEntry, final Eva
...
+
+        public FastCachingMapVariableResolverFactory(final Map variables) {
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java]: Seems can be private

File: src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java [hunk 3]
```diff
@@ -88,38 +94,71 @@ public Serializable evaluateOutputEntry(final OutputEntry outputEntry, final Eva
...
+            if(vr == null) {
+                addResolver(name, vr = new SimpleSTValueResolver(value, null, true));
+            } else {
+                vr.setValue(value);
+            }
+
+            return vr;
+        }
+
+        public VariableResolver getVariableResolverOrNull(String name) {
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java]: Seems can be private

File: src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java [hunk 4]
```diff
@@ -88,38 +94,71 @@ public Serializable evaluateOutputEntry(final OutputEntry outputEntry, final Eva
...
+            }
+            else if (variables.containsKey(name)) {
+                variableResolvers.put(name, vr = new SimpleSTValueResolver(variables.get(name), null));
+                return vr;
+            }
+            else if (nextFactory != null) {
+                return nextFactory.getVariableResolver(name);
+            }
+
+            return null;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java]: What about to limit returns to one?

**Comment by mariuszkumor** [comment]
Moreover, I can see sonar violations https://sonarcloud.io/project/issues?id=org.powerflows%3Admn&pullRequest=202&resolved=false

And don't you think, any unit test testing a proper cache usage should be done in this PR?

**Comment by stlhrt** [comment]
Cache is tinternal, imho there is no need to test beyond contract.

**Pull request merged**